source/multigrid/mg_transfer_internal.cc

```diff
@@ -131,42 +143,67 @@ namespace internal
-                    {
-                      copy_indices_global_mine[level].emplace_back(
-                        global_dof_indices[i], level_dof_indices[i]);
-
-                      // send this to the owner of the level_dof:
-                      send_data_temp.emplace_back(level,
-                                                  global_dof_indices[i],
-                                                  level_dof_indices[i]);
+                      // we own both the active dof index and the level one ->
+                      // set them into the vector, indexed by the local index
+                      // range of the active dof
+                      unrolled_copy_indices[mg_dof.locally_owned_dofs()
+                                              .index_within_set(
+                                                global_dof_indices[i])] =
+                        level_dof_indices[i];
                     }
                   else
                     {
-                      // somebody will send those to me
-                    }
+                      // now we collect them in the pair
+                      const unsigned int relevant_idx =
+                        globally_relevant.index_within_set(
+                          global_dof_indices[i]);
+
+                      // skip if we did this global dof already (on this or a
+                      // coarser level)
+                      if (dof_touched[relevant_idx])
+                        continue;
+
+                      if (global_mine)
+                        {
+                          copy_indices_global_mine[level].emplace_back(
+                            global_dof_indices[i], level_dof_indices[i]);
+
+                          // send this to the owner of the level_dof:
+                          send_data_temp.emplace_back(level,
+                                                      global_dof_indices[i],
+                                                      level_dof_indices[i]);
+                        }
+                      else
+                        {
+                          // somebody will send those to me
+                        }
 
-                  dof_touched[global_idx] = true;
+                      dof_touched[relevant_idx] = true;
+                    }
                 }
             }
+
+          // we now translate the plain vector for the copy_indices field into
+          // the expected format of a pair of indices.
+          if (has_active_cells_on_this_level)
+            {
+              copy_indices[level] = {};
+              if (copy_indices_global_mine[level].empty())
+                copy_indices[level].reserve(unrolled_copy_indices.size());
+              for (unsigned int i = 0; i < unrolled_copy_indices.size(); ++i)
+                if (unrolled_copy_indices[i] != numbers::invalid_dof_index)
+                  copy_indices[level].emplace_back(
+                    mg_dof.locally_owned_dofs().nth_index_in_set(i),
```

Comment:
So `nth_index_in_set` here and `index_within_set` above are not overly expensive?

Reply:
No, because they are on locally owned dofs; I added comments.